provider/marathon/marathon.go

```diff
@@ -420,6 +447,187 @@ func (p *Provider) hasCircuitBreakerLabels(application marathon.Application) boo
 	return ok
 }
 
+// Map of services properties
+// we can get it with label[serviceName][propertyName] and we got the propertyValue
+type labelServiceProperties map[string]map[string]string
+
+// Check if for the given task and applications, we find labels that are defining services
+func (p *Provider) hasTaskServices(task marathon.Task, applications []marathon.Application) bool {
+	application, err := getApplication(task, applications)
+	if err != nil {
+		log.Errorf("Unable to get marathon application from task %s", task.AppID)
+		return false
+	}
+	return len(extractServicesLabels(application.Labels)) > 0
+}
+
+// Check if for the given application, we find labels that are defining services
+func (p *Provider) hasServices(application marathon.Application) bool {
+	return len(extractServicesLabels(application.Labels)) > 0
+}
+
+// Extract the service labels from application labels of dockerData struct
+func extractServicesLabels(labels *map[string]string) labelServiceProperties {
+	v := make(labelServiceProperties)
+
+	for index, serviceProperty := range *labels {
+		matches := servicesPropertiesRegexp.FindStringSubmatch(index)
+		if matches != nil {
+			result := make(map[string]string)
+			for i, name := range servicesPropertiesRegexp.SubexpNames() {
+				if i != 0 {
+					result[name] = matches[i]
+				}
+			}
+			serviceName := result["service_name"]
+			if _, ok := v[serviceName]; !ok {
+				v[serviceName] = make(map[string]string)
+			}
+			v[serviceName][result["property_name"]] = serviceProperty
+		}
+	}
+
+	return v
```

Comment:
Thanks for the explanation. My main concern is about how the regular expression is applied in the code, and how the regexp package API is used to extract the kind of information we are looking for. To be honest, I think the API isn't the most intuitive, which may stem from the fact that regular expressions aren't the most trivial topic.

Let me share my thoughts on your issues with the parser approach:

1. **code size.** Yes, we'd have somewhat more code. IMO though, I think it'd be easier to understand. When I tried to understand your regex, I basically had to jump back and forth between the regex and the algorithm in `extractServicesLabels` to convince myself that the code is really covering all cases. Personally, I tend to prefer a bit less dense code at the benefit of better comprehensibility.
1. **parser logic/testing.** For sure we'd need a custom parser. I disagree on the testing argument though: The regex and the parser approaches should have roughly the same test cases. In fact, I think the more explicit nature of the parser approach make it much clearer to see which cases need to be tested. On the contrary, it's harder to figure out the circumstances under which a regular expression matches or not (which is what took me so long to understand).
3. **performance.** Regexes require full-blown state machines whereas parsing is mostly a matter of simple string comparisons and manipulations. So I have my doubts whether the regex will outperform the parser. More importantly though, I wouldn't worry about this matter at all: It probably takes a lot of regex'ing/parsing for any performance impact to be observable, and even if we think it was, we'd need benchmarks to prove it. For now, let's focus on a solution that's easiest to read, comprehend, and test.

I'll add another point in favor of parsers: they are often simpler to extend. When the structure of what your regular expression has to match changes, chances are you'll have to rebuild the expression or start introducing hard-to-understand features like look-ahead/look-behind. That's usually where things get ugly.

FWIW, the stdlib is full of parsers, so at least the Go team thinks it's a good and fast-enough thing. :)

One implementation speaks louder than 100 discussions, so I went ahead and [implemented the parser](https://github.com/timoreimann/traefik/commit/bdab1458dd0a25ace8b8775d964fca304897407c) for comparison reasons. I admit it's a bit larger (especially because your implementation can be further simplified now that we say the service name is mandatory -- I'll address that in a separate comment). OTOH, it's super clear to see what to expect from the label pattern when being parsed, which makes writing down all test cases a rather mechanical matter.

Overall, I'm leaning towards the parser for clarity reasons, but I can see the advantage of a shorter implementation when using a regular expression. I guess I can convince myself that staying with what we have right now in the PR is acceptable, though I wouldn't want to leave this comment thread without a rather complete discussion. :-)

I'll also try to get some additional feedback from the other maintainers on this point once we've covered the remainders of the PR.